File: autoeditor.js
/* mpv integration with auto-editor (https://github.com/WyattBlue/auto-editor)
 *
 * Changelog:
 * https://github.com/idMysteries/mpv-skip-silence/commits
 * == v3 (2022/03/28)
 * Made the script work with the new version of auto-editor
 * == v2 (2022/02/01)
 * The script can now invoke auto-editor automatically with a keybind.
 *
 * Limitations:
 * 1. The video must have a constant frame-rate; variable frame rate sources,
 *    sources with frame skips will have issues.
 *
 * Copyright 2022 Tatsuyuki Ishi <ishitatsuyuki@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

var AUTO_EDITOR_BIN = "auto-editor";
var AUTO_EDITOR_ARGS = ["--export_as_json", "-mcut", "60", "-t", "0.04"];
var SILENCE_SPEED = 3;

var in_silence = false;
var restore_speed = 1.0;

var timeObserver;

var cmdInProgress = false;

function runAutoEditor() {
	if (cmdInProgress) {
		mp.osd_message("auto-editor: An analysis is already in progress");
		return;
	}
	var file = mp.get_property("path");
	var cmd = {
		name: "subprocess",
		playback_only: false,
		args: [AUTO_EDITOR_BIN, file].concat(AUTO_EDITOR_ARGS)
	};
	mp.osd_message("auto-editor: Running analysis");
	cmdInProgress = true;
	mp.command_native_async(cmd, function (success, result, error) {
		cmdInProgress = false;
		if (success) {
			load();
		} else {
			mp.msg.error(error);
		}
	});
}

function load() {
	unload(); // Unload previous callbacks
	var file = mp.get_property("path");
	file = file.replace(/\.[^.]+$/, ".json");
	var content;
	try {
		content = JSON.parse(mp.utils.read_file(file));
	} catch (e) {
		return;
	}
  
	var segments = content["chunks"]
	if (segments.length < 1) return;
  
	mp.osd_message(
		"auto-editor: Loaded " + segments.length + " segments"
	);
  
	var current_segment = segments[0];
  
	timeObserver = function (_name, time) {
		var frame = mp.get_property_number("estimated-frame-number");
	
		if (frame >= current_segment[0] && frame < current_segment[1]) {
			if (current_segment[2] === 99999) {
				if (in_silence === false){
					in_silence = true;
					restore_speed = mp.get_property_number("speed");
					mp.set_property("speed", SILENCE_SPEED);
				}
			}
			else {
				if (in_silence){
					in_silence = false;
					mp.set_property("speed", restore_speed);
				}
			}
		}
		else {
			for (var i = 0; i < segments.length; i++) {
				if (frame < segments[i][1]) {
					current_segment = segments[i];
					break;
				}
			}
		}
	};
	mp.observe_property("time-pos", "number", timeObserver);
}

function unload() {
	if (timeObserver != null) {
		mp.unobserve_property(timeObserver);
		timeObserver = null;
	}
}

mp.register_event("start-file", load);
mp.register_event("end-file", unload);
mp.add_key_binding("E", "run-auto-editor", runAutoEditor);

File: autoeditor.lua
-- MPV integration with auto-editor (https://github.com/WyattBlue/auto-editor)

local utils = require "mp.utils"
local msg = require "mp.msg"
local options = require "mp.options"

local AUTO_EDITOR_BIN = "auto-editor"

local config = {
    enabled = false,
    restore_speed = 1.0,
    silence_speed = 2.5,
    threshold = "4%",
    margin = "0.1s,0.2s",
    mincut = "1s",
}
options.read_options(config)

local time_observer = nil
local cmd_in_progress = false

local function update_playback_speed(speed)
    mp.set_property("speed", speed)
    msg.info("Speed set to: " .. speed)
end

local function is_local_file(file_path)
    if not file_path then
        return false
    end
    local file_info = utils.file_info(file_path)
    return file_info ~= nil and file_info.is_file
end

local function load_segments(json)
    if time_observer then
        mp.unobserve_property(time_observer)
        time_observer = nil
        update_playback_speed(config.restore_speed)
    end

    local parsed_content = utils.parse_json(json)
    if not parsed_content then
        msg.error("Failed to parse JSON.")
        return
    end

    local segments = parsed_content["chunks"]
    if not segments or #segments < 1 then
        msg.warn("No segments found.")
        return
    end

    mp.osd_message("Loaded " .. #segments .. " segments")
    msg.info("Loaded " .. #segments .. " segments")

    local current_segment = nil
    time_observer = function(_, time)
        local frame = mp.get_property_number("estimated-frame-number") or 0
        if current_segment and frame >= current_segment[1] and frame < current_segment[2] then
            return
        end
        for _, segment in ipairs(segments) do
            if frame >= segment[1] and frame < segment[2] then
                current_segment = segment
                if segment[3] == 99999 then
                    local current_speed = mp.get_property_number("speed")
                    if current_speed < config.silence_speed then
                        config.restore_speed = current_speed
                        update_playback_speed(config.silence_speed)
                    end
                else
                    update_playback_speed(config.restore_speed)
                end
                return
            end
        end
        current_segment = nil
    end

    mp.observe_property("time-pos", "number", time_observer)
end

local function execute_auto_editor()
    if cmd_in_progress then
        mp.osd_message("Analysis is already in progress")
        msg.info("Analysis is already in progress")
        return
    end

    local file = mp.get_property("path")
    if not is_local_file(file) then
        mp.osd_message("auto-editor: disabled for network streams")
        msg.info("Disabled for network streams")
        return
    end

    local audio_stream_index = (mp.get_property_number("aid") or 1) - 1

    local auto_editor_args = {
        file,
        "--export", "timeline:api=1",
        "--quiet",
        "--no-cache",
        "--progress", "none",
        "--margin", config.margin,
        "--edit", string.format("audio:stream=%d,threshold=%s,mincut=%s", audio_stream_index, config.threshold, config.mincut)
    }

    local cmd = {
        name = "subprocess",
        playback_only = false,
        capture_stdout = true,
        args = {AUTO_EDITOR_BIN, unpack(auto_editor_args)}
    }

    mp.osd_message("Running analysis")
    msg.info("Running analysis")
    cmd_in_progress = true

    mp.command_native_async(cmd, function(success, result, error)
        cmd_in_progress = false
        if success then
            load_segments(result.stdout)
        else
            local err_msg = error or "Unknown error"
            msg.error("Error: " .. err_msg)
            mp.osd_message("Error: " .. err_msg)
        end
    end)
end

local function auto_start_analysis()
    if config.enabled then
        if is_local_file(mp.get_property("path")) then
            execute_auto_editor()
        else
            mp.osd_message("auto-editor: disabled for network streams")
            msg.info("Disabled for network streams")
        end
    end
end

local function display_settings()
    local settings_str = string.format(
        "Auto-editor settings:\n" ..
        "Auto-run: %s\n" ..
        "Restore speed: %.2f\n" ..
        "Silence speed: %.2f\n" ..
        "Threshold: %s\n" ..
        "Margin: %s",
        tostring(config.enabled),
        config.restore_speed,
        config.silence_speed,
        config.threshold,
        config.margin
    )
    mp.osd_message(settings_str, 5)
    msg.info(settings_str)
end

mp.add_key_binding("E", "run-auto-editor", execute_auto_editor)
mp.add_key_binding("Ctrl+E", "print-auto-editor-settings", display_settings)
mp.register_event("file-loaded", auto_start_analysis)
